File: 2016/Day7/main.cpp
#include <fstream>
#include <iostream>
#include <regex>
#include <string>
#include <unordered_set>
#include <vector>

bool contains_abba(const std::string& s) {
    for (size_t i = 0; i < s.size() - 3; ++i) {
        if (s[i] != s[i + 1] && s[i] == s[i + 3] && s[i + 1] == s[i + 2]) {
            return true;
        }
    }
    return false;
}

std::unordered_set<std::string> collect_aba(const std::string& s) {
    std::unordered_set<std::string> aba;
    for (size_t i = 0; i < s.size() - 2; ++i) {
        if (s[i] != s[i + 1] && s[i] == s[i + 2]) {
            aba.insert(s.substr(i, 3));
        }
    }
    return aba;
}

bool contains_bab(const std::string& s, const std::unordered_set<std::string>& abas) {
    for (size_t i = 0; i < s.size() - 2; ++i) {
        if (s[i] != s[i + 1] && s[i] == s[i + 2]) {
            std::string bab = {s[i + 1], s[i], s[i + 1]};
            if (abas.find(bab) != abas.end()) {
                return true;
            }
        }
    }
    return false;
}

void split_segments(const std::string& ip, std::vector<std::string>& supernet, std::vector<std::string>& hypernet) {
    std::regex re(R"((\[|\]))");
    std::regex_token_iterator it(ip.begin(), ip.end(), re, {-1, 0});
    bool inside_brackets = false;
    for (; it != std::sregex_token_iterator(); ++it) {
        if (it->str() == "[") {
            inside_brackets = true;
        } else if (it->str() == "]") {
            inside_brackets = false;
        } else {
            if (inside_brackets) {
                hypernet.push_back(it->str());
            } else {
                supernet.push_back(it->str());
            }
        }
    }
}

int main() {
    std::ifstream input_file("input.txt");
    std::vector<std::string> ips;
    std::string line;

    while (std::getline(input_file, line)) {
        ips.push_back(line);
    }

    int tls_count = 0;
    int ssl_count = 0;

    for (const auto& ip : ips) {
        std::vector<std::string> supernet;
        std::vector<std::string> hypernet;
        split_segments(ip, supernet, hypernet);

        bool has_abba_outside = false;
        bool has_abba_inside = false;

        for (const auto& segment : supernet) {
            if (contains_abba(segment)) {
                has_abba_outside = true;
                break;
            }
        }

        for (const auto& segment : hypernet) {
            if (contains_abba(segment)) {
                has_abba_inside = true;
                break;
            }
        }

        if (has_abba_outside && !has_abba_inside) {
            tls_count++;
        }

        std::unordered_set<std::string> abas;
        for (const auto& segment : supernet) {
            auto segment_abas = collect_aba(segment);
            abas.insert(segment_abas.begin(), segment_abas.end());
        }

        bool supports_ssl = false;
        for (const auto& segment : hypernet) {
            if (contains_bab(segment, abas)) {
                supports_ssl = true;
                break;
            }
        }

        if (supports_ssl) {
            ssl_count++;
        }
    }

    std::cout << "Part 1: " << tls_count << std::endl;
    std::cout << "Part 2: " << ssl_count << std::endl;

    return 0;
}

File: 2016/Day7/main.lua
local function contains_abba(s)
    for i = 1, #s - 3 do
        if s:sub(i, i) ~= s:sub(i + 1, i + 1) and
           s:sub(i, i) == s:sub(i + 3, i + 3) and
           s:sub(i + 1, i + 1) == s:sub(i + 2, i + 2) then
            return true
        end
    end
    return false
end

local function collect_aba(s)
    local aba = {}
    for i = 1, #s - 2 do
        if s:sub(i, i) ~= s:sub(i + 1, i + 1) and
           s:sub(i, i) == s:sub(i + 2, i + 2) then
            table.insert(aba, s:sub(i, i + 2))
        end
    end
    return aba
end

local function contains_bab(s, abas)
    for i = 1, #s - 2 do
        if s:sub(i, i) ~= s:sub(i + 1, i + 1) and
           s:sub(i, i) == s:sub(i + 2, i + 2) then
           local bab = s:sub(i + 1, i + 1) .. s:sub(i, i) .. s:sub(i + 1, i + 1)
           for _, aba in ipairs(abas) do
               if aba == bab then
                   return true
               end
           end
        end
    end
    return false
end

local function split_segments(ip, supernet, hypernet)
    local inside_brackets = false
    local last_index = 1

    for i = 1, #ip do
        local char = ip:sub(i, i)
        if char == '[' then
            if i > last_index then
                table.insert(supernet, ip:sub(last_index, i - 1))
            end
            inside_brackets = true
            last_index = i + 1
        elseif char == ']' then
            if i > last_index then
                table.insert(hypernet, ip:sub(last_index, i - 1))
            end
            inside_brackets = false
            last_index = i + 1
        end
    end

    if last_index <= #ip then
        local segment = ip:sub(last_index)
        if inside_brackets then
            table.insert(hypernet, segment)
        else
            table.insert(supernet, segment)
        end
    end
end

local ips = {}
for line in io.lines('input.txt') do
    table.insert(ips, line)
end

local tls_count = 0
local ssl_count = 0

for _, ip in ipairs(ips) do
    local supernet = {}
    local hypernet = {}
    split_segments(ip, supernet, hypernet)

    local has_abba_outside = false
    local has_abba_inside = false

    for _, segment in ipairs(supernet) do
        if contains_abba(segment) then
            has_abba_outside = true
            break
        end
    end

    for _, segment in ipairs(hypernet) do
        if contains_abba(segment) then
            has_abba_inside = true
            break
        end
    end

    if has_abba_outside and not has_abba_inside then
        tls_count = tls_count + 1
    end

    local abas = {}
    for _, segment in ipairs(supernet) do
        for _, aba in ipairs(collect_aba(segment)) do
            table.insert(abas, aba)
        end
    end

    local supports_ssl = false
    for _, segment in ipairs(hypernet) do
        if contains_bab(segment, abas) then
            supports_ssl = true
            break
        end
    end

    if supports_ssl then
        ssl_count = ssl_count + 1
    end
end

print('Part 1: ' .. tls_count)
print('Part 2: ' .. ssl_count)
